numalogic/registry/dynamodb_registry.py

```diff
@@ -0,0 +1,398 @@
+        self.__save_item(
+            table, part_key, self._construct_version_key(sort_key), artifact_data, version=version
+        )
+
+        # Delete the stale model
+        stale_version = version - self.models_to_retain
+        if stale_version > 0:
+            _LOGGER.info("Deleting stale version %s", stale_version)
+            self.__delete_item(
+                table, part_key, self._construct_version_key(sort_key, version=stale_version)
+            )
+
+        return str(version)
+
+    @staticmethod
+    def __delete_item(table, part_key: str, v_sort_key: str):
+        try:
+            response = table.delete_item(Key={"skey": part_key, "dkey": v_sort_key})
+        except ClientError as err:
+            raise DynamoDBRegistryError(f"{err.__class__.__name__} raised") from err
+        return response
+
+    def delete(self, skeys: KEYS, dkeys: KEYS, version: str) -> Optional[dict[str, Any]]:
+        """
+        Deletes the artifact from dynamodb.
+        Args:
+            skeys: static key fields as list/tuple of strings
+            dkeys: dynamic key fields as list/tuple of strings
+            version: the version of the artifact to be deleted.
+        """
+        if version == "0":
+            raise ValueError("Cannot delete the main (zeroth) record directly!")
+
+        table = self.connector.get_resource().Table(self.table_name)
+        part_key = self.construct_key(skeys)
+        sort_key = self.construct_key(dkeys)
+
+        artifact_data = self.load(skeys, dkeys, latest=True)
+
+        if artifact_data:
+            # Check if the latest version is getting deleted
+            if version == artifact_data.extras.get("version"):
+                # Find if an older version exists
+                prev_version = str(int(version) - 1)
+                artifact_data_prev = self.load(skeys, dkeys, latest=False, version=prev_version)
+                if artifact_data_prev and prev_version != "0":
+                    # Update the zeroth record pointer to the previous version
+                    prev_item = self._pack_artifact_data(
+                        artifact_data_prev.artifact, prev_version, **artifact_data_prev.metadata
+                    )
+                    self.__save_item(
+                        table,
+                        part_key,
+                        self._construct_version_key(sort_key),
+                        prev_item,
+                        version=prev_version,
+                    )
+                else:
+                    # Delete the zeroth record pointer if no previous version exists
+                    self.__delete_item(table, part_key, self._construct_version_key(sort_key))
+
+        # Delete the main version record
+        return self.__delete_item(table, part_key, self._construct_version_key(sort_key, version))
+
+    @staticmethod
+    def _construct_version_key(sort_key: str, version: int = 0) -> str:
```

Comment:
Do you think it is possible to have 'latest' as the pointer instead of v0?

Reply:
The rational to use a number instead of switching between numbers and strings is that we can always convert to int, then increase, decrease, compare, order etc. In addition, this is an internal implementation and DynamoDB pattern, should not be exposed through any of the public methods.

Reply:
When do you think we would call a particular version instead of just the latest in real time?